### ISSUE TITLE
Config container mandatory when containing only mandatory "config false" decendants 

### ISSUE THREAD

**Issue by kwatsen** [CZ-NIC/yangson#124]
```
yangson yl.json -v config.json
Schema error: {/} missing-data: expected 'ietf-netconf-acm:nacm'
```

Where yl.json is:
```
{      
  "ietf-yang-library:modules-state": {
    "module-set-id": "",
    "module": [
      {
        "name": "ietf-yang-types",
        "revision": "2013-07-15",
        "namespace": "urn:ietf:params:xml:ns:yang:ietf-yang-types",
        "conformance-type": "import"
      },
      {
        "name": "ietf-netconf-acm",
        "revision": "2018-02-14",
        "namespace": "urn:ietf:params:xml:ns:yang:ietf-netconf-acm",
        "conformance-type": "implement"
      }
    ]
  }
}
```

And config.json is:
```
{
}
```

This issue effectively reopens #33.   I agree that containers become mandatory when descendants are mandatory, but in this case, the mandatory descendants are "config false" and so should not apply when validating config.

FWIW, `yanglint` exhibits the expected behavior (not requiring `/ietf-netconf-acm:nacm` when only validating config):
```
# VALIDATES AS "CONFIG"
$ yanglint -t config ietf-netconf-acm@2018-02-14.yang config.json
<no output>

# FAILS AS "DATA"
$ yanglint -t data ietf-netconf-acm@2018-02-14.yang config.json
libyang err : Mandatory node "denied-operations" instance does not exist. (Schema location "/ietf-netconf-acm:nacm/denied-operations".)
YANGLINT[E]: Failed to parse input data file "config.json".
```

**Comment by llhotka** [comment]
Strictly speaking, such a container is mandatory according to [section 3](https://datatracker.ietf.org/doc/html/rfc7950#section-3) of RFC 7950:

> *  A container node without a "presence" statement and that has at least one mandatory node as a child.

I agree though that it is not very practical when validating config data, and the issue with NACM pops up in many places. So I am going to check how difficult it would be to handle this.

**Comment by kwatsen** [comment]
Any update on this?

It would be nice to remove some special work-around code I have...

**Comment by llhotka** [comment]
No, not yet. It is not exactly straightforward, more substantial changes would be needed.

**Comment by llhotka** [comment]
As a workaround, it might help to change `conformance-type` for `ietf-netconf-acm` to `import`. In my experience, in most modules NACM is only present through things like
```
nacm:default-deny-write;
```
and `import` suffices for this.

**Comment by kwatsen** [comment]
When using NACM for its flags, I always do just “import” and life is good. 

In this case, I’m connecting to  YumaPro server and Andy has NACM as an “implemented” built-in, even though it’s not being used.

FWIW there are nearly 25 built-in modules, thankfully none other have a mandatory config and so can be ignored. 

By the way, Yuma itself is okay with the parent container being empty in config.   In my code, I only create-delete the empty container before/after calling Yangson ‘validate’.

**Comment by kwatsen** [comment]
This issue has now escaped the "NACM box".  That is, I have now run across this issue in the wild.  Specifically, vendor-created module having a `config true`  node with a mix of CT and CF descendants, but some of the CF descendants are `mandatory true`.

Validating the response from `<get-config>`  throws this exception (pro-tip: I replaced the specific path with `example-module:top-level-node`):

```
/Users/kent/.pyenv/versions/3.11.4/lib/python3.11/site-packages/yangson/schemanode.py:610: in _validate
    self._check_schema_pattern(inst, ctype)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <yangson.schemanode.SchemaTreeNode object at 0x102e6c210>, inst = <yangson.instance.RootNode object at 0x105e2ff50>
ctype = <ContentType.config: 1>

    def _check_schema_pattern(self: "InternalNode", inst: InstanceNode,
                              ctype: ContentType) -> None:
        p = self.schema_pattern
        p._eval_when(inst)
        for m in inst:
            newp = p.deriv(m, ctype)
            if isinstance(newp, NotAllowed):
                raise SchemaError(
                    inst,
                    ("" if ctype == ContentType.all else ctype.name + " ") +
                    "member-not-allowed", m)
            p = newp
        if not p.nullable(ctype):
            mms = p._mandatory_members(ctype)
            msg = "one of " if len(mms) > 1 else ""
>           raise SchemaError(
                inst, "missing-data",
                "expected " + msg + ", ".join([repr(m) for m in mms]))
E           yangson.exceptions.SchemaError: {/} missing-data: expected 'example-module:top-level-node'

yangson/schemanode.py:638: SchemaError
```

FWIW, the vendor's server is implemented using `Yuma Pro`.   I only mention this because Yuma allows this response to be returned, meaning Yangon is in the rough.

**Comment by kwatsen** [comment]
You wrote above "It is not exactly straightforward, more substantial changes would be needed."

Can you provide more detail?  I may be able to help...

**Comment by llhotka** [comment]
The mandatory property for internal nodes (containers in particular) is determined upfront based on “mandatoriness” of the child nodes, and permeated up in the schema tree. Corresponding schema patterns are then stored with each `InternalNode`. For example, consider this module:
```
module issue124 {
  namespace "http://example.com/issue124";
  prefix i124;

  container top {
    container ambo {
      leaf config {
        type uint8;
      }
      leaf nonconfig {
        config false;
        mandatory true;
        type uint8;
      }
    }
  }
}
```
The schema pattern for the `ambo` container is then simply
```
Member ambo
```
which means that it has to be present. After removing the `mandatory true;` statement, the schema pattern changes to
```
Alternative
  Empty
  Member ambo
```
I don't want to have separate schema patterns for config and nonconfig data, but we can try other options. One solution might be to introduce a new pattern like `EmptyInConfig` and change the `mandatory` property of `SchemaNode` from simple boolean to something else in order to be able to represent “mandatory unless in config”.

**Comment by kwatsen** [comment]
FWIW, the stacktrace for the pytest update is:

```
yangson/instance.py:420: in validate
    self.schema_node._validate(self, scope, ctype)
yangson/schemanode.py:632: in _validate
    inst._member(m).validate(scope, ctype)
yangson/instance.py:420: in validate
    self.schema_node._validate(self, scope, ctype)
yangson/schemanode.py:966: in _validate
    super()._validate(inst, scope, ctype)
yangson/schemanode.py:630: in _validate
    self._check_schema_pattern(inst, ctype)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <yangson.schemanode.ContainerNode object at 0x10478bfd0>, inst = <yangson.instance.ObjectMember object at 0x1046ca110>
ctype = <ContentType.all: 3>

    def _check_schema_pattern(self: "InternalNode", inst: InstanceNode,
                              ctype: ContentType) -> None:
        p = self.schema_pattern
        p._eval_when(inst)
        for m in inst:
            newp = p.deriv(m, ctype)
            if isinstance(newp, NotAllowed):
                raise SchemaError(
                    inst,
                    ("" if ctype == ContentType.all else ctype.name + " ") +
                    "member-not-allowed", m)
            p = newp
        if not p.nullable(ctype):
            mms = p._mandatory_members(ctype)
            msg = "one of " if len(mms) > 1 else ""
>           raise SchemaError(
                inst, "missing-data",
                "expected " + msg + ", ".join([repr(m) for m in mms]))
E           yangson.exceptions.SchemaError: {/test:contA} missing-data: expected 'issue124'
```

**Comment by kwatsen** [comment]
Can you describe how `schema_pattern` works?  I'm trying to reverse-engineer it, but am getting weird behavior.  For example, if a CF container has two MT leafs, the `schema_pattern` is always just the 2nd/last leaf, but I was expecting both leafs, since both are MT.  Sometimes I see, e.g. "Empty or member FooBar", which I read "if not empty, FooBar must exist", but that doesn't seem right either...

Regarding:

> One solution might be to introduce a new pattern like EmptyInConfig and _**change the mandatory property of SchemaNode from simple boolean to something else**_ in order to be able to represent “mandatory unless in config”.

I'm super-leery of the `mandatory` property diverging from YANG.  Anything else invites great confusion, and also would be a NBC change.  IMO, the solution should be 100% internal.

> I don't want to have separate schema patterns for config and nonconfig data, but we can try other options. 

Having separate patterns was my first thought.  Can you say why it doesn't appeal to you?

**Comment by llhotka** [comment]
It's based on [this algorithm](https://relaxng.org/jclark/derivative.html), with a few extensions that were needed for YANG. I am looking into it, hopefully I can have something by the end of the next week.

I didn't mean to change the meaning of `mandatory` in YANG, just in the Python code.

**Comment by kwatsen** [comment]
That algorithm hurts my head  ;)

Thank you for looking into it!

**Comment by kwatsen** [comment]
Thank you for digging in.

Is my test case helpful?  Would it be good for me prepare a more exhaustive test case, to cover different scenarios of manditoriness (e.g., nested containers, when, etc.)?

Enjoy your vacation!

**Comment by llhotka** [comment]
What to do with this corner case that is IMO legal?
```
container bag {
    choice either {
        mandatory true;
        leaf foo {
            config true;
            ...
        }
        leaf bar {
            config false;
            ...
        }
    }
}
```
Is `bag` mandatory in configuration or not?

**Comment by kwatsen** [comment]
I read it that ‘bag’ is always present.  

This is an unusual structure.   I’m having a hard time visualizing when it might occur in practice.  In config, “foo” would have to be set.   In opstate, either “foo” or “bar” would have to be set but, since “foo” must be set in config, it (NMDA?) implies that “foo” (not “bar”) would be set in opstate.

**Comment by llhotka** [comment]
I agree it is weird. So let's keep the current approach: `bag` is mandatory no matter  what.

**Comment by llhotka** [comment]
OK, the `mandstate` branch now handles the above example correctly and all tests pass. So please check it on your use case.

**Comment by kwatsen** [comment]
Nice!

The branch passes most of my tests (and definitely fixes the NACM-issue), but I'm having an issue with a container, having only mandatory CF leafs, inside a container, inside a container.   `inst.validate()` is saying that one of the higher-up containers is expected (`missing-data`).   If I edit the YANG module commenting out the `mandatory true` lines for the CF leafs, validation succeeds.

**Comment by llhotka** [comment]
Right, the properties didn't propagate correctly up the tree. Please test the new version in `mandstate` branch.

**Comment by kwatsen** [comment]
The current `mandstate` branch passes all my tests.  I did some extra tests, which passed, and there are some more I'd like to do, but right now everything looks great!

At one point I found the new code falling into a forever loop, but I haven't been already to reproduce it...and lost the original stack trace  :(

**Comment by llhotka** [comment]
Hmm, it would be great if you can reproduce the infinite loop. In the mean time, I am going to update documentation and tests.

**Comment by kwatsen** [comment]
I wrote the "more" tests mentioned in my previous comment and everything performed very well.

I'm unable to reproduce the infinite loop.  This is unfortunate, but since everything appears to be working, maybe it's okay to merge?   It'll get broader testing in the wild, but then again, "mandatory true" CF nodes are rare...

**Comment by llhotka** [comment]
OK, will do.

**Comment by llhotka** [comment]
It's included in version 1.4.19.
